gateway/src/apicast/policy/default_credentials/apicast-policy.json

```diff
@@ -0,0 +1,21 @@
+{
+  "$schema": "http://apicast.io/policy-v1/schema#manifest#",
+  "name": "Default credentials",
+  "summary": "Provides default credentials for unauthenticated requests",
+  "description":
+    ["This policy allows to expose a service without authentication. ",
+     "It can be useful, for example, for legacy apps that cannot be adapted to ",
+     "send the auth params. ",
+     "When the credentials are not provided in the request, this policy ",
+     "provides the default ones configured. ",
+     "An app_id + app_key or a user_key should be configured."],
+  "version": "builtin",
+  "configuration": {
+    "type": "object",
+    "properties": {
+      "user_key": { "type": "string" },
```

Comment:
I was hoping this actually could be "oneOf" and user would chose user_key/app_id. But that probably can't happen because of the limitation of the react library?

Reply:
I didn't find a way to do that.
Something like this does not work:
```json
{  
   "type":"object",
   "properties":{  
      "credentials":{  
         "type":"object",
         "oneOf":[  
            {  
               "type":"object",
               "properties":{  
                  "user_key":{  
                     "type":"string"
                  }
               }
            },
            {  
               "type":"object",
               "properties":{  
                  "app_key":{  
                     "type":"string"
                  },
                  "app_id":{  
                     "type":"string"
                  }
               }
            }
         ]
      }
   }
}
```

Reply:
Thats exactly what I had in mind. 

What about:

```json
{
  "type": "object",
  "properties": {
    "credentials": {
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "name": {
            "type": "string",
            "enum": ["app_id", "app_key", "user_key"]
          },
          "value": {
            "type": "string"
          }
        }
      }
    }
  }
}
```

Reply:
That can be loaded in the react lib, but it has other problems.

With that solution, an array with app_id, app_key and user_key is valid. Also, an array with user_key and app_key for example, or an array with only app_key.

Also, with that solution including, for example, 2 different user_keys is valid.

Reply:
If we keep the backend version detection, then it works exactly the same as current solution.
It has exactly the same problems as it is now: you can fill any combination of them.

We can make one field required if other field is filled in: https://github.com/mozilla-services/react-jsonschema-form#property-dependencies (but that would not work with this array).

I think with https://github.com/mozilla-services/react-jsonschema-form#dynamic we could do a selector to chose a backend version and then let them fill app_id+app_key / user_key.

I'm not advocating for actually doing it, but I'd like to evaluate our options. 
Also I think we don't really have to strongly validate everything. If they want to fill in just app_key, let them. 


BTW what if we want to expand it to support JWT/OAuth/TLS in the future? Would one structure be better over another? Without `oneOf` not really, right?

Reply:
`dependencies` would fail to validate things like incluing an app_id with a user_key.
The dynamic schema would work. You can try it in https://mozilla-services.github.io/react-jsonschema-form/ to see how it looks:
```json
{
    "type":"object",
    "properties":{
      "auth_type":{
        "type":"string",
        "enum":[
          "user_key",
          "app_id_and_app_key"
        ],
        "default":"user_key"
      }
    },
    "required":[
      "auth_type"
    ],
    "dependencies":{
      "auth_type":{
        "oneOf":[
          {
            "properties":{
              "auth_type":{
                "enum":[
                  "user_key"
                ]
              },
              "user_key":{
                "type":"string"
              }
            },
            "required":[
              "user_key"
            ]
          },
          {
            "properties":{
              "auth_type":{
                "enum":[
                  "app_id_and_app_key"
                ]
              },
              "app_id":{
                "type":"string"
              },
              "app_key":{
                "type":"string"
              }
            },
            "required":[
              "app_id",
              "app_key"
            ]
          }
        ]
      }
    }
  }
```

With that schema, the config would be something like:
```json
{
    "auth_type": "user_key",
    "user_key": "some_user_key"
}
```

or

```json
{
  "auth_type": "app_id_and_app_key",
  "app_id": "some_app_id",
  "app_key": "some_app_key"
}
```

It looks good to me. What do you think?

Reply:
That looks good to me!

Reply:
Changed :+1: